api/src/main/java/edu/wpi/first/shuffleboard/api/data/DataTypes.java

```diff
@@ -86,6 +61,104 @@ public static void register(DataType<?> dataType) {
   }
 
   /**
+   * Gets the data type most relevant to a Java class.
+   */
+  @SuppressWarnings("unchecked")
+  public static <T> Optional<DataType<T>> forJavaType(Class<T> type) {
+    if (type.isPrimitive()) {
+      return forJavaType((Class) boxedType(type));
+    }
+    return (Optional) typeCache.computeIfAbsent(type, __ -> {
+      if (DataType.class.isAssignableFrom(type)) {
+        return forType((Class<DataType>) type);
+      }
+      Optional<DataType> naive = dataTypes.values().stream()
+          .filter(t -> type.equals(t.getJavaClass()))
+          .findAny();
+      if (naive.isPresent()) {
+        return naive;
+      } else {
+        // Check Java class hierarchy
+        Comparator<Class<?>> classComparator = closestTo(type);
+        List<DataType> sorted = dataTypes.values().stream()
+            .filter(t -> t != All)
+            .sorted((t1, t2) -> classComparator.reversed().compare(t1.getJavaClass(), t2.getJavaClass()))
+            .collect(Collectors.toList());
+        return Optional.of(sorted.get(0));
+      }
+    });
+  }
+
+  private static Comparator<Class<?>> closestTo(Class<?> target) {
```

Comment:
Can you explain why this is needed and what it is? Also, this is a complex bit of logic, are you sure that Guava doesn't have some sort of reflection utility to do this?
https://github.com/google/guava/wiki/ReflectionExplained

Also, if you do need it, I really hope that this is unit tested.